pandas/core/reshape.py

```diff
@@ -558,87 +549,138 @@ def _convert_level_number(level_num, columns):
             else:
                 return columns.names[level_num]
 
+    if isinstance(level_nums, int):
+        level_nums = [level_nums]
+
     this = frame.copy()
 
     # this makes life much simpler
-    if level_num != frame.columns.nlevels - 1:
-        # roll levels to put selected level at end
-        roll_columns = this.columns
-        for i in range(level_num, frame.columns.nlevels - 1):
+    # roll levels to put selected level(s) at end
+    roll_columns = this.columns
+    for j, level_num in enumerate(reversed(level_nums)):
+        for i in range(level_num, frame.columns.nlevels - (j + 1)):
             # Need to check if the ints conflict with level names
             lev1 = _convert_level_number(i, roll_columns)
             lev2 = _convert_level_number(i + 1, roll_columns)
             roll_columns = roll_columns.swaplevel(lev1, lev2)
-        this.columns = roll_columns
+    this.columns = roll_columns
 
     if not this.columns.is_lexsorted():
         # Workaround the edge case where 0 is one of the column names,
-        # which interferes with trying to sort based on the first
-        # level
+        # which interferes with trying to sort based on the first level
         level_to_sort = _convert_level_number(0, this.columns)
         this = this.sortlevel(level_to_sort, axis=1)
 
-    # tuple list excluding level for grouping columns
-    if len(frame.columns.levels) > 2:
+    num_levels_to_stack = len(level_nums)
+    level_vals = this.columns.levels[-num_levels_to_stack:]
+    level_labels = sorted(set(zip(*this.columns.labels[-num_levels_to_stack:])))
+    level_vals_used = MultiIndex.from_tuples([tuple(np.nan if lab == -1 else level_vals[i][lab]
```

Comment:
I'm figuring out which combinations of the values to be stacked (the final num_levels_to_stack levels in this.columns) are actually used. This is to avoid introducing extra rows corresponding to combinations of levels that don't actually have any columns (i.e. when the column levels aren't a complete product).